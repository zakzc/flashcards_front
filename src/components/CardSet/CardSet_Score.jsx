import React, { Component } from "react";

export default class Score extends Component {
  render() {
    return (
      <React.Fragment>
        <div className="scoreArea">
          <div id="ongoingScore">
            Right answers:{" "}
            <div className="scoreNumbers">{this.props.rightAnswers}</div>
            Wrong answers:{" "}
            <div className="scoreNumbers">{this.props.wrongAnswers}</div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}
