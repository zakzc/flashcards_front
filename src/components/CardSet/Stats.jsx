import React, { Component } from "react";
// hooks
// import { useDB_Connection } from "../../DB/DB-hook/connection-hook";
import RewindIcon from "../icons/rewind";
import StatsIcon from "../icons/stats";

export default class Stats extends Component {
  constructor(props) {
    super(props);
    this.state = {
      right: props.right,
      wrong: props.wrong,
      lengthOfStack: props.lengthOfStack,
      guesses: props.right + props.wrong,
      percentage: ((props.right / (props.right + props.wrong)) * 100).toFixed(
        2
      ),
    };
    this.reSetGame = props.reSetGame;
    this.reStart = this.reStart.bind(this);
    this.stackLevel = this.stackLevel.bind(this);
    // this.pieChart = this.pieChart.bind(this);
    this.reStartButton = this.reStartButton.bind(this);
    this.gameStatsButton = this.gameStatsButton.bind(this);
    this.gameStatsContents = this.gameStatsContents.bind(this);
    this.statsLayout = this.statsLayout.bind(this);
  }

  ////////////////
  // Implementation methods
  ////////////////

  reStart(event) {
    event.preventDefault();
    // TODO: nodeJs routing to save the data into json (log)??
    // window.location.reload(false);
    this.reSetGame();
    console.log("current card: ", this.state.current);
  }

  // pieChart() {
  //   let correct = parseInt(this.state.percentage, 10) + 0.53;
  //   console.log("correct: ", correct, typeof correct);
  //   let cover = correct.toString() + " 100";
  //   console.log("cover: ", cover, typeof cover);
  //   return (
  //     <div id="pieChartGraph">
  //       {/* <svg id="pieGraph" viewBox="0 0 32 32" transform="rotate(270)">
  //         <circle className="first" strokeDasharray={cover}></circle>
  //       </svg> */}
  //       <svg width="100" height="100" id="pieGraph">
  //         <circle r="25" cx="50" cy="50" class="pie" strokeDasharray={cover} />
  //       </svg>
  //     </div>
  //   );
  // }

  stackLevel() {
    let level = "";
    let infoLevel = "";
    if (this.state.lengthOfStack <= 10) {
      level = " easy";
      infoLevel = " (any stack with less than 10 cards is considered easy).";
    } else if (this.state.lengthOfStack <= 20) {
      level = " in an intermediate level";
      infoLevel =
        " (any stack with more than 10 cards, but less than 20, is considered easy).";
    } else {
      level = " hard";
      infoLevel = " (any stack with more than 20 cards is considered hard).";
    }
    return (
      <div>
        <p>
          {" "}
          This is considered to be
          <strong className="highlight">{level}</strong>
          {infoLevel}
        </p>
      </div>
    );
  }

  ////////////////
  // View methods
  ////////////////

  gameStatsButton() {
    return (
      <button className="buttonStyle" type="button" onClick={this.reStart}>
        <div className="flipContainer">
          <div className="flipInner">
            <div className="flipFront">
              <StatsIcon className="buttonStyle" />
            </div>
            <div className="flipBack">
              <span className="buttonMessage">Game statistics</span>
            </div>{" "}
          </div>
        </div>
      </button>
    );
  }

  gameStatsContents() {
    return (
      <div id="gameStatsContents">
        <h2>Game Stats</h2>
        <div>
          You finished your stack. This stack was made up of
          <strong className="highlight"> {this.state.lengthOfStack} </strong>
          cards. {this.stackLevel()}
        </div>
        <p>
          You guessed{" "}
          <strong className="highlight"> {this.state.guesses}</strong> times, of
          which
          <strong className="highlight"> {this.state.wrong} </strong> were
          wrong.
        </p>
        <div id="percentage">
          That means you got{" "}
          <strong className="highlight"> {this.state.percentage}% </strong>of
          your guesses right.
          <br />
          {/* <div id="graphView">
            Graphical view (<strong className="statEmphasis">yellow </strong>{" "}
            represents the right answers):
          </div> */}
          {/* {this.pieChart()} */}
        </div>
      </div>
    );
  }

  gameStats() {
    return (
      <React.Fragment>
        <div>
          {this.gameStatsButton()}
          {this.gameStatsContents()}
        </div>
      </React.Fragment>
    );
  }

  reStartButton() {
    return (
      <button className="buttonStyle" type="button" onClick={this.reStart}>
        <div className="flipContainer">
          <div className="flipInner">
            <div className="flipFront">
              <RewindIcon className="buttonStyle" />
            </div>
            <div className="flipBack">
              <span className="buttonMessage"> Start over</span>
            </div>
          </div>
        </div>
      </button>
    );
  }

  reStartTheGame() {
    return (
      <div>
        {this.reStartButton()}
        <h4 id="reStartQuestion">Do you want to re-start the stack?</h4>
      </div>
    );
  }

  //TODO make a an array with the cards that were wrong in CardSet and present them in the stats

  statsLayout() {
    return (
      <div id="gameOverWrap">
        <div id="gameStats">{this.gameStats()}</div>
        <div id="reStartMe">{this.reStartTheGame()}</div>
      </div>
    );
  }

  render() {
    return <React.Fragment>{this.statsLayout()}</React.Fragment>;
  }
}
