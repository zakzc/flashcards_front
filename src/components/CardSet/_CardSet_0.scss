//////////////
// * General Layout
//////////////

#cardText {
  grid-area: cardText;
}

#cardFront {
  background-color: $mainColor;
  padding: 100px 200px 100px 200px;
  border-radius: 20px;
  font-size: 100%;
}

#cardBack {
  background-color: $contrastColor;
  padding: 100px 200px 100px 200px;
  border-radius: 20px;
  font-size: 100%;
}

#card {
  grid-area: card;
  display: grid;
  justify-content: center;
  align-content: center;
  background-color: $mainBackColor;
}

#info {
  grid-area: info;
}

// #buttonsArea {
//   grid-area: buttonsArea;
//   display: grid;
//   background-color: green;
//   grid-template-areas: "flip control control";
// }

#scoreCardArea {
  grid-area: scoreCardArea;
}

#cardArea {
  grid-area: cardArea;
  display: grid;
  grid-template-areas: "card";
  grid-template-rows: 3fr 1fr;
}

#controlCardArea {
  grid-area: cardControl;
  display: grid;
  grid-template-areas:
    "buttonsArea"
    "scoreCardArea"
    "info";
  grid-template-rows: repeat(3, auto-fit);
  gap: 15px;
}

#flipButton {
  grid-area: flip;
  background-color: $mainBackColor;
  margin-top: 25px;
}

#controlButtons {
  grid-area: control;
}

#buttonsArea {
  display: grid;
  grid-template-areas: "flip control control";
  background-color: $mainBackColor;
}

#buttonsWrapper {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  margin-top: 25px;
}

//////////////
// * Responsive
//////////////

@media only screen and (max-width: 767px) {
  #cardSetLayout {
    display: grid;
    grid-template-areas:
      "cardArea"
      "cardControl";
    grid-template-columns: 2.5fr 1.5fr;
  }
}

// bigger screen sizes > 768
@media only screen and (min-width: 768px) {
  #cardSetLayout {
    display: grid;
    grid-template-areas: "cardArea cardControl";
    grid-template-columns: 2.5fr 1.5fr;
    // gap: 20px;
  }
}
