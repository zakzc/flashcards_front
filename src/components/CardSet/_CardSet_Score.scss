// elements that apply to all screen sizes:
#scoreCardArea {
  padding: 1%;
}
