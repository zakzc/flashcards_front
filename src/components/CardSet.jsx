import React from "react";
import { Redirect } from "react-router-dom";
// import { TransitionGroup, Transition } from "react-transition-group";
import { AnimatePresence, motion } from "framer-motion";
// Standard Log out page
import UserIsLoggedOut from "./views/UserIsLoggedOut";
// module
import Stats from "./CardSet/Stats";
import Score from "./CardSet/Score";
// icons
import Check from "./icons/check";
import Right from "./icons/right";
import Wrong from "./icons/wrong";

export default class CardSet extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userIsLoggedIn: props.userIsLoggedIn,
      currentStack: props.currentStack,
      cards: props.currentStack.cards,
      current: 0,
      face: true,
      right: 0,
      wrong: 0,
      lengthOfStack: props.currentStack.cards.length,
      stackIsOver: false,
    };
    // external modules
    this.UserIsLoggedOut = UserIsLoggedOut;
    this.Stats = Stats;
    this.Score = Score;
    // icons
    // this.turning = turning;
    // methods associated with rendering
    this.cardSide = this.cardSide.bind(this);
    this.cardCount = this.cardCount.bind(this);
    this.buttonArea = this.buttonArea.bind(this);
    this.controlButtons = this.controlButtons.bind(this);
    this.flipButton = this.flipButton.bind(this);
    /*     this.score = this.score.bind(this); */
    // methods for child comp
    this.reSetGame = this.reSetGame.bind(this);
    // methods associated with cards
    this.handleClick = this.handleClick.bind(this);
    this.removeFromPile = this.removeFromPile.bind(this);
    this.setAnswerRight = this.setAnswerRight.bind(this);
    this.setAnswerWrong = this.setAnswerWrong.bind(this);
    this.nextCardAvailable = this.nextCardAvailable.bind(this);
    this.updateCurrentCard = this.updateCurrentCard.bind(this);
    this.playCardSet = this.playCardSet.bind(this);
    this.cardSetLayout = this.cardSetLayout.bind(this);
    this.cardSetView = this.cardSetView.bind(this);
    this.infoCard = this.infoCard.bind(this);
  }

  // Reload method, will be called from Stats
  reSetGame() {
    this.setState({
      userIsLoggedIn: this.props.userIsLoggedIn,
      currentStack: this.props.currentStack,
      cards: this.props.currentStack.cards,
      current: 0,
      face: true,
      right: 0,
      wrong: 0,
      lengthOfStack: this.props.currentStack.cards.length,
      stackIsOver: false,
    });
  }

  ////////////////
  // Implementation methods
  ////////////////

  // Removes the item from the pile (called upon Right answer)
  removeFromPile() {
    // Most checks are done in the nextCard() method, however, some need to
    // be performed here or there will be errors on render().
    const newCardSet = this.state.cards.slice();
    newCardSet.splice(this.state.current, 1);
    this.setState(() => ({
      cards: newCardSet,
    }));
    this.updateCurrentCard();
  }

  // These functions set answerIsRight/wrong in the card array of objects to true/false,
  // Then, it calls the next functions depending on the conditions/
  setAnswerRight() {
    this.setState(() => ({
      face: true,
      right: this.state.right + 1,
    }));
    // No need to increment, on removing, the array will set back in
    // we need, though, to check if on removing, the pile won't collapse
    if (this.nextCardAvailable(this.state)) {
      this.removeFromPile(this.state);
    } else {
      // if, on removing, the pile is empty, the stack is over
      this.setState({
        stackIsOver: true,
      });
    }
    return;
  }

  setAnswerWrong() {
    this.setState(() => ({
      face: true,
      wrong: this.state.wrong + 1,
    }));
    if (this.nextCardAvailable(this.state)) {
      this.updateCurrentCard();
    }
    return;
  }

  // Sequence of logical checks to determine if we should call a next card or not
  nextCardAvailable() {
    // Is the pile empty?
    if (this.state.cards.length === 0) {
      this.setState(() => ({ stackIsOver: true }));
      return false;
    } else {
      return true;
    }
  }

  // Updates the number of the current cards
  updateCurrentCard() {
    // key aspect is increment or decrement current, to keep the
    // app from looking for a card number that doesn't exist
    let next = this.state.current + 1;
    if (next >= this.state.cards.length - 1) {
      next = 0;
    }
    this.setState(() => ({
      current: next,
    }));
    return;
  }

  // * Page rendering functions

  // This function shows the back of the current card (works like a switch button)
  handleClick() {
    this.setState((state) => ({
      face: !state.face,
    }));
  }

  ////////////////
  // View methods
  ////////////////

  infoCard() {
    return (
      <div id="shortMessage" className="infoCard_cardSet">
        <p>
          Flip the card with the flip button. Then, indicate ✓ for right answer
          or ✘ for wrong answer.
        </p>
      </div>
    );
  }

  cardSide() {
    try {
      if (this.state.cards.length === 0 || this.state.stackIsOver) {
        this.setState({ stackIsOver: true });
      } else {
        return (
          <React.Fragment>
            <motion.div
              /*     transition={{ ease: "easeIn", duration: 0.7 }} */
              className="cardText"
            >
              {this.state.face ? (
                <AnimatePresence>
                  <motion.div
                    whileHover={{ scale: 1.02 }}
                    /*               initial={{ opacity: 0 }}
                  animate={{ opacity: 1 }} */
                    /*    exit={{ rotateX: 360 }} */
                    id="cardFront"
                  >
                    {this.state.cards[this.state.current].front}
                  </motion.div>
                </AnimatePresence>
              ) : (
                /*            <AnimatePresence> */
                <motion.div
                  whileHover={{ scale: 1.02 }}
                  animate={{ rotateY: 360 }}
                  transition={{ ease: "easeInOut", duration: 0.7 }}
                  id="cardBack"
                >
                  {this.state.cards[this.state.current].back}
                </motion.div>
                /*               </AnimatePresence> */
              )}
            </motion.div>
          </React.Fragment>
        );
      }
    } catch (E) {
      console.log("error on rendering", E);
      console.log(
        "Review state:",
        this.state,
        "length: ",
        this.state.cards.length
      );
    }
  }

  flipButton() {
    return (
      <div id="flipButton_Positioning">
        <div className="flipContainer">
          <div className="flipInner">
            <button
              className="buttonStyle"
              type="button"
              onClick={this.handleClick}
            >
              <div className="flipFront">
                <Check />
              </div>
              <div className="flipBack">
                <span className="buttonMessage">Flip</span>
              </div>
            </button>
          </div>
        </div>
      </div>
    );
  }

  rightAnswerButton() {
    return (
      <div id="correctButton_Positioning">
        <div className="flipContainer">
          <div className="flipInner">
            <button
              className="buttonStyle"
              type="button"
              onClick={() => this.setAnswerRight(this.state)}
            >
              {" "}
              <div onClick={this.handleClick}>
                <div className="flipFront">
                  <Right />
                </div>
                <div className="flipBack">
                  <span className="buttonMessage">Right</span>
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    );
  }

  wrongAnswerButton() {
    return (
      <div id="wrongButton_Positioning">
        <div className="flipContainer">
          <div className="flipInner">
            <button
              className="buttonStyle"
              type="button"
              onClick={() => this.setAnswerWrong(this.state)}
            >
              <div onClick={this.handleClick}>
                <div className="flipFront">
                  <Wrong />
                </div>
                <div className="flipBack">
                  <span className="buttonMessage">Wrong</span>
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    );
  }

  // rendering of right or wrong buttons and call for the setCard
  controlButtons() {
    // only shows when the card is in the back view (face===false)
    if (!this.state.face)
      return (
        <React.Fragment>
          <div id="buttonsWrapper">
            {this.rightAnswerButton()}
            {this.wrongAnswerButton()}
          </div>
        </React.Fragment>
      );
  }

  buttonArea() {
    return (
      <div id="buttonsArea">
        <div id="flipButton">{this.flipButton()}</div>
        <div id="controlButtons">{this.controlButtons()}</div>
      </div>
    );
  }

  cardCount() {
    if (
      this.state.cards.length !== 0 ||
      (this.state.current !== 0 && this.state.cards.length !== 0)
    ) {
      return (
        <React.Fragment>
          <div className="scoreArea">
            <div id="currentCard">
              Card number:
              <div className="scoreNumbers">
                <p>
                  {this.state.current + 1}/{this.state.cards.length}
                </p>
              </div>
            </div>
            <br />
          </div>
        </React.Fragment>
      );
    }
  }

  playCardSet() {
    return (
      <React.Fragment>
        <div id="cardArea">
          <div id="card">{this.cardSide()}</div>
        </div>
        <div id="controlCardArea">
          {this.buttonArea()}
          <div id="info">{this.infoCard()}</div>
          <div id="scoreCardArea">
            <Score
              rightAnswers={this.state.right}
              wrongAnswers={this.state.wrong}
            />
            <div id="cardCount"> {this.cardCount()} </div>

            <br />
            <div id="currentPlayStack">
              {" "}
              <p>Current stack is: </p>
              <strong className="currentSelectedStack">
                {" "}
                {this.state.currentStack.stackName}
              </strong>
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }

  cardSetLayout() {
    return (
      <div id="cardSetLayout">
        {this.state.stackIsOver ? (
          <Stats
            right={this.state.right}
            wrong={this.state.wrong}
            lengthOfStack={this.state.lengthOfStack}
            reSetGame={this.reSetGame}
          />
        ) : (
          <div id="cardAreaWrapper">{this.playCardSet()}</div>
        )}
      </div>
    );
  }

  cardSetView() {
    console.log("CARD SET PLAY");
    return (
      <div id="cardSetView">
        {this.state.userIsLoggedIn ? (
          <div id="loggedInView">{this.cardSetLayout()}</div>
        ) : (
          <Redirect to="/" />
        )}
      </div>
    );
  }

  render() {
    return <React.Fragment>{this.cardSetView()}</React.Fragment>;
  }
}
