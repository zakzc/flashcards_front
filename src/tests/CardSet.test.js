import React from "react";
import Enzyme from "enzyme";
///
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
///
import CardSet from "../components/CardSet";
import Stats from "../components/CardSet/Stats";

Enzyme.configure({ adapter: new Adapter() });

// Parent
describe("\n __| CardSet |__", () => {
  const cardSet = shallow(<CardSet />);

  describe("Initialization process", () => {
    it("Renders the CardSet component", () => {
      expect(cardSet).toMatchSnapshot();
    });

    it("Initializes the State", () => {
      expect(cardSet.state().face).toEqual(true);
      expect(cardSet.state().right).toEqual(0);
      expect(cardSet.state().wrong).toEqual(0);
      expect(cardSet.state().stackIsOver).toEqual(false);
    });

    describe("Imports", () => {
      test("UserIsLoggedOut", () => {
        expect(cardSet.find("UserIsLoggedOut")).toBeTruthy();
      });
      test("Stats", () => {
        expect(cardSet.find("Stats")).toBeTruthy();
      });
      test("Score", () => {
        expect(cardSet.find("Score")).toBeTruthy();
      });
      describe("Svg imports", () => {
        test("Check", () => {
          expect(cardSet.find("Check")).toBeTruthy();
        });
        test("Right", () => {
          expect(cardSet.find("Right")).toBeTruthy();
        });
        test("Wrong", () => {
          expect(cardSet.find("Wrong")).toBeTruthy();
        });
      });
    });
  });
});

// Child
describe("\n- -> | Stats (child) | ", () => {
  const stats = shallow(<Stats />);

  describe("\tInitialization process", () => {
    it("Renders the ManageCards component", () => {
      expect(stats).toMatchSnapshot();
    });

    describe("\tImports", () => {
      describe("\tSvg imports", () => {
        test("RewindIcon", () => {
          expect(stats.find("RewindIcon")).toBeTruthy();
        });
        test("StatsIcon", () => {
          expect(stats.find("StatsIcon")).toBeTruthy();
        });
      });
    });
  });
});
