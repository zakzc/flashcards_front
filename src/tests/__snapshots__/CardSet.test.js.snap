// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`
 __| CardSet |__ Initialization process Renders the CardSet component 1`] = `ShallowWrapper {}`;

exports[`
- -> | Stats (child) |  	Initialization process Renders the ManageCards component 1`] = `ShallowWrapper {}`;
